net/sched/sch_gkprioq.c

```diff
@@ -0,0 +1,395 @@
+
+}
+
+static struct sk_buff *gkprioq_dequeue(struct Qdisc *sch)
+{
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+	struct sk_buff *skb;
+	int err;
+	
+
+	skb = qdisc_dequeue_head(q->prio_array[q->highest_prio]);
+
+	if (skb) {
+		if (q->last_dropped_prio != -1) {
+			err = fifo_set_limit(q->prio_array[q->highest_prio], 
+				   (q->prio_array[q->highest_prio])->limit - 1);
+			if (err)
+				pr_warn_once("Something is wrong!");
+			
+			/* Donate resources to priority queue whose last packet 			 * was dropped because of resource unavailability then.
+	                 */
+
+			err = fifo_set_limit(q->prio_array[q->last_dropped_prio]			     ,(q->prio_array[q->last_dropped_prio])->limit + 1);
+
+			if (err)
+				pr_warn_once("Something is wrong!");
+	
+		}
+
+		goto done; 
+	}
+
+	return NULL;
+
+done:
+	if ((q->prio_array[q->highest_prio])->q.qlen > 0)
+		return skb;
+
+	q->highest_prio = calc_new_high_prio(q); 
+	return skb;
+
+}
+
+static int gkprioq_init(struct Qdisc *sch, struct nlattr *opt)
+{
+	int prio;
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+	struct Qdisc *pfifo_q;
+
+	if (opt == NULL) 
+		 q->max_limit = qdisc_dev(sch)->tx_queue_len; 
+	else {
+		struct tc_gkprioq_qopt *ctl = nla_data(opt);
+
+		if(nla_len(opt) < sizeof(*ctl))
+			return -EINVAL;
+
+		q->max_limit = ctl->limit;
+	}
+	
+	q->init_limit = 1;
+
+	for(prio = 0; prio <= 1; prio++) {
+		pfifo_q = fifo_create_dflt(sch, &pfifo_qdisc_ops, 
+					q->max_limit/(MAX_PRIORITY + 1));
+		if (IS_ERR(pfifo_q)) 
+			return PTR_ERR(pfifo_q);
+		q->prio_array[prio] = pfifo_q;
+	}
+
+	for(prio = 2; prio < MAX_PRIORITY + 1; prio++) {
+		pfifo_q = fifo_create_dflt(sch, &pfifo_qdisc_ops, 
+						q->init_limit);
+		if (IS_ERR(pfifo_q)) 
+			return PTR_ERR(pfifo_q);
+		q->prio_array[prio] = pfifo_q;
+	}
+
+	/* Saving resources now and deploying on the fly allocation in order to 	 * avoid under utilization of available resources
+	 */
+
+	q->last_dropped_prio = -1;
+	q->highest_prio = 0; 
+	return 0;
+
+}
+
+static int gkprioq_dump(struct Qdisc *sch, struct sk_buff *skb)
+{
+	struct tc_gkprioq_qopt opt = {.limit = sch->limit};
+	
+	if (nla_put(skb, TCA_OPTIONS, sizeof(opt), &opt))
+		goto nla_put_failure;
+	return skb->len;
+
+nla_put_failure:
+	return -1;
+}
+
+static void gkprioq_reset(struct Qdisc *sch)
+{
+	int prio;
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+
+	for (prio = 0; prio < MAX_PRIORITY + 1; prio++) 
+		qdisc_reset(q->prio_array[prio]);	
+
+	sch->qstats.backlog = 0;
+	q->init_limit = 1;
+	q->last_dropped_prio = -1;
+	q->highest_prio = 0;
+}
+
+static void gkprioq_destroy(struct Qdisc *sch)
+{
+	int prio;
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+
+	for (prio = 0; prio < MAX_PRIORITY + 1; prio++)
+		qdisc_destroy(q->prio_array[prio]);
+}
+
+struct Qdisc_ops gkprioq_qdisc_ops __read_mostly = {
+	.id		=	"gkprioq",
+	.priv_size	=	sizeof(struct gkprioq_sched_data),
+	.enqueue	=	gkprioq_enqueue,
+	.dequeue	=	gkprioq_dequeue,
+	.peek		=	qdisc_peek_dequeued,
```

Comment:
This is correct because gk queues may change the head of the queue when a higher priority packet comes in between a peek and a dequeue, and code that calls `peek()` expects the head of the queue to be constant. `qdisc_peek_dequeued()` deals with these cases. It's weird, but it's the interface.